src/allmydata/cli/grid_manager.py

```diff
@@ -0,0 +1,213 @@
+    node-directory. NAME is an arbitrary label.
+    """
+    try:
+        ctx.obj.grid_manager.add_storage_server(
+            name,
+            ed25519.verifying_key_from_string(public_key),
+        )
+    except KeyError:
+        raise click.ClickException(
+            "A storage-server called '{}' already exists".format(name)
+        )
+    save_grid_manager(
+        _config_path_from_option(ctx.parent.params["config"]),
+        ctx.obj.grid_manager,
+        create=False,
+    )
+    return 0
+
+
+@grid_manager.command()
+@click.argument("name")
+@click.pass_context
+def remove(ctx, name):
+    """
+    Remove an existing storage-server by name from a Grid Manager
+    """
+    fp = _config_path_from_option(ctx.parent.params["config"])
+    try:
+        ctx.obj.grid_manager.remove_storage_server(name)
+    except KeyError:
+        raise click.ClickException(
+            "No storage-server called '{}' exists".format(name)
+        )
+    cert_count = 0
+    if fp is not None:
+        while fp.child('{}.cert.{}'.format(name, cert_count)).exists():
+            fp.child('{}.cert.{}'.format(name, cert_count)).remove()
+            cert_count += 1
+
+    save_grid_manager(fp, ctx.obj.grid_manager, create=False)
+
+
+@grid_manager.command()
+@click.pass_context
+def list(ctx):
+    """
+    List all storage-servers known to a Grid Manager
+    """
+    for name in sorted(ctx.obj.grid_manager.storage_servers.keys()):
+        blank_name = " " * len(name)
+        click.echo("{}: {}".format(name, ctx.obj.grid_manager.storage_servers[name].public_key_string()))
+        for cert in ctx.obj.grid_manager.storage_servers[name].certificates:
+            delta = datetime.utcnow() - cert.expires
+            click.echo("{}  cert {}: ".format(blank_name, cert.index), nl=False)
+            if delta.total_seconds() < 0:
+                click.echo("valid until {} ({})".format(cert.expires, abbreviate_time(delta)))
+            else:
+                click.echo("expired {} ({})".format(cert.expires, abbreviate_time(delta)))
+
+
+@grid_manager.command()
+@click.argument("name")
+@click.argument(
+    "expiry_days",
+    type=click.IntRange(1, 5*365),  # XXX is 5 years a good maximum?
+)
+@click.pass_context
+def sign(ctx, name, expiry_days):
+    """
+    sign a new certificate
+    """
+    fp = _config_path_from_option(ctx.parent.params["config"])
+    expiry = timedelta(days=expiry_days)
+
+    try:
+        certificate = ctx.obj.grid_manager.sign(name, expiry)
+    except KeyError:
+        raise click.ClickException(
+            "No storage-server called '{}' exists".format(name)
+        )
+
+    certificate_data = json.dumps(certificate, indent=4)
+    click.echo(certificate_data)
+    if fp is not None:
+        next_serial = 0
+        f = None
+        while f is None:
+            try:
+                f = fp.child("{}.cert.{}".format(name, next_serial)).create()
+            except Exception:
```

Comment:
This looks a little broad.

Reply:
okay, I changed in to `OSError` but I don't know how to tell what exceptions might come from `create()` (especially, e.g. on non-linux platforms).